app-typescript/components/Dropdown.tsx

```diff
@@ -0,0 +1,270 @@
+import * as React from 'react';
+import classNames from 'classnames';
+
+export interface IMenuItem {
+    label: string;
+    icon?: string;
+    onSelect(): void;
+}
+
+export interface ISubmenu {
+    type: 'submenu';
+    label: string;
+    icon?: string;
+    items: Array<IMenuItem | ISubmenu | IMenuGroup | 'divider'>;
+}
+
+export interface IMenuGroup {
+    type: 'group';
+    label?: string;
+    items: Array<IMenuItem | ISubmenu | IMenuGroup | 'divider'>;
+}
+
+interface IMenu {
+    label?: string;
+    align?: 'left' | 'right';
+    items: Array<IMenuItem | ISubmenu | IMenuGroup | 'divider'>;
+    header?: Array<IMenuItem | ISubmenu | IMenuGroup | 'divider'>;
+    footer?: Array<IMenuItem | ISubmenu | IMenuGroup | 'divider'>;
+    children: React.ReactNode;
+}
+
+export const Dropdown = ({
+    items,
+    header,
+    footer,
+    children,
+    align,
+}: IMenu) => {
+    const [open, setOpen] = React.useState(false);
+    const [change, setChange] = React.useState(false);
+    const [height, setHeight] = React.useState(false);
+    const [submenu, setSubmenu] = React.useState(false);
+    const [width, setWidth] = React.useState(false);
+    const ref = React.useRef(null);
+    let inDebounce = 0;
+
+    React.useLayoutEffect(() => {
+        let element = document.getElementsByClassName('dropdown')[0];
+        let parentElement = getScrollParent(element);
+        function applyScroll() {
+            return change ? debounce(50) : null;
+        }
+        parentElement.parentNode.addEventListener("scroll", applyScroll());
+
+        calculate();
+
+        return () => {
+            parentElement.removeEventListener("scroll", applyScroll());
+            clearTimeout(inDebounce);
+        };
+    }, [open]);
+
+    React.useEffect(() => {
+        setChange(true);
+    }, [open]);
+
+    // open close
+    function isOpen() {
```

Comment:
A better name would help. Maybe `toggleDisplay` or the same `setOpen` you are already using. You could change the existing one to `setOpenState` for example to avoid collisions. is + adjective naming is mostly used to read variables and return a boolean.